backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java

```diff
@@ -99,45 +109,86 @@ public LiveOktaRepository(
 
     // By default, when creating a user, we give them privileges of a standard user
     String organizationExternalId = org.getExternalId();
-    Set<OrganizationRole> roles = EnumSet.of(OrganizationRole.getDefault(), role);
-    Set<String> groupIds = new HashSet<>();
-
-    for (OrganizationRole r : roles) {
-      String groupName = generateGroupName(organizationExternalId, r);
-
-      // Okta SDK's way of getting a group by group name
-      GroupList groups = _client.listGroups(groupName, null, null);
-      if (groups.stream().count() == 0) {
-        throw new IllegalGraphqlArgumentException(
-            String.format("Cannot add Okta user to nonexistent group=%s", groupName));
-      }
-      Group group = groups.single();
-      groupIds.add(group.getId());
+    Set<OrganizationRole> rolesToCreate = EnumSet.of(OrganizationRole.getDefault());
+    rolesToCreate.addAll(roles);
+
+    // Add user to new groups
+    Set<String> groupNamesToAdd = new HashSet<>();
+    groupNamesToAdd.addAll(
+        rolesToCreate.stream()
+            .map(r -> generateRoleGroupName(organizationExternalId, r))
+            .collect(Collectors.toSet()));
+    groupNamesToAdd.addAll(
+        facilities.stream()
+            // use an empty set of facilities if user can access all facilities anyway
+            .filter(f -> !PermissionHolder.grantsAllFacilityAccess(rolesToCreate))
+            .map(f -> generateFacilityGroupName(organizationExternalId, f.getInternalId()))
+            .collect(Collectors.toSet()));
+
+    GroupList orgGroups =
+        _client.listGroups(generateGroupOrgPrefix(organizationExternalId), null, null);
+    if (orgGroups.stream().count() == 0) {
+      throw new IllegalGraphqlArgumentException(
+          String.format(
+              "Cannot add Okta user to nonexistent organization=%s", organizationExternalId));
     }
+    Set<String> orgGroupNames =
+        orgGroups.stream().map(g -> g.getProfile().getName()).collect(Collectors.toSet());
+    groupNamesToAdd.stream()
+        .filter(n -> !orgGroupNames.contains(n))
```

Comment:
So this seems like if it ever arises we're going to get a phone call in a hurry, since by the time we get here all the group names have been validated as things that _should_ existt. I'm a little concerned that this will blow up on us at some point if we add a new role, because at that point we'll need to do an Okta migration that could have been taken care of with on-demand creation. That said, that's not a today problem.